### PR TITLE
32-bit Mac Support

### PR DESCRIPTION
**Pull request by RPangrle**
Ugh. Legacy ObjC runtime.

On top of adding manual memory management (that's all those SAFE_ARC_* macros) I also had to move all ivar declarations to headers.

@synthesize is required, as properties don't synthesize themselves on the legacy runtime.

I removed some abuse of properties as well, as they caused some very peculiar behavior, particularly around memory management.

### CODE REVIEW

**Comment by RPangrle** [comment]
@kpanwar You might want to peek at this.

**Comment by kpanwar** [comment]
@RPangrle - I assume you are looking at the travis issues and this is not simply a build failure?

**Comment by RPangrle** [comment]
Already committed a fix, I tried pushing it yesterday but the wifi on the bus wasn't so great.

File: ADAL/src/ADAuthenticationContext.m [hunk 1]
```diff
@@ -118,10 +131,12 @@ - (id)initWithAuthority:(NSString *)authority
     ADTokenCache* cache = [ADTokenCache new];
     [cache setDelegate:delegate];
     
-    return [self initWithAuthority:authority
+    self = [self initWithAuthority:authority
                  validateAuthority:validateAuthority
                         tokenCache:cache
                              error:error];
+    SAFE_ARC_RELEASE(cache);
+    return self;
```
Comment (jasoncoolmax) [inline ADAL/src/ADAuthenticationContext.m]: Why we don't need to SAFE_ARC_AUTORELEASE(self) before return?

File: ADAL/src/ADClientMetrics.m [hunk 2]
```diff
@@ -50,31 +50,38 @@ + (ADClientMetrics*) getInstance {
...
         _endpoint = endPoint;
+        SAFE_ARC_RETAIN(_endpoint);
+        SAFE_ARC_RELEASE(_responseTime);
         _responseTime = @"";
+        SAFE_ARC_RETAIN(_responseTime);
+        SAFE_ARC_RELEASE(_correlationId);
         _correlationId = correlationId;
+        SAFE_ARC_RETAIN(_correlationId);
+        SAFE_ARC_RELEASE(_startTime);
         _startTime = [NSDate new];
```
Comment (jasoncoolmax) [inline ADAL/src/ADClientMetrics.m]: Why we don't need to retain _startTime?

File: ADAL/src/ADInstanceDiscovery.m [hunk 3]
```diff
@@ -149,9 +136,11 @@ -(void) validateAuthority: (NSString*) authority
     NSString* message = [NSString stringWithFormat:@"Attempting to validate the authority: %@; CorrelationId: %@", authority, [correlationId UUIDString]];
     AD_LOG_VERBOSE(@"Instance discovery", correlationId, message);
     
-    ADAuthenticationError* error;
+    authority = [authority lowercaseString];
```
Comment (jasoncoolmax) [inline ADAL/src/ADInstanceDiscovery.m]: Is there a memory leak here? :P

File: ADAL/src/cache/ADTokenCacheItem.m [hunk 4]
```diff
@@ -48,55 +47,71 @@ - (void)calculateHash
...
+    item->_authority = [_authority copyWithZone:zone];
+    item->_clientId = [_clientId copyWithZone:zone];
+	item->_familyId = [_familyId copyWithZone:zone];
+    item->_accessToken = [_accessToken copyWithZone:zone];
+    item->_accessTokenType = [_accessTokenType copyWithZone:zone];
+    item->_refreshToken = [_refreshToken copyWithZone:zone];
+    item->_expiresOn = [_expiresOn copyWithZone:zone];
+    item->_userInformation = [_userInformation copyWithZone:zone];
+    item->_sessionKey = [_sessionKey copyWithZone:zone];
+	item->_tombstone = [_tombstone mutableCopyWithZone:zone];
```
Comment (jasoncoolmax) [inline ADAL/src/cache/ADTokenCacheItem.m]: if _tombstone is nil, we want the copy value also be nil, so should it be 
item->_tombstone = _tombstone ? [_tombstone mutableCopyWithZone:zone] : nil;
?

File: ADAL/src/urlprotocol/ADCustomHeaderHandler.m [hunk 5]
```diff
@@ -20,42 +20,47 @@
...
-        customHeaders = @{
-                         @"x-ms-PkeyAuth":@"1.0",
-                         };
-        customHeadersForSingleUse = [NSMutableDictionary new];
+    if (self == [ADCustomHeaderHandler class])
+    {
+        s_customHeaders =
+            [NSMutableDictionary dictionaryWithDictionary: @{@"x-ms-PkeyAuth":@"1.0"}];
+        SAFE_ARC_RETAIN(s_customHeaders);
+        s_customHeadersForSingleUse = [NSMutableDictionary new];
```
Comment (jasoncoolmax) [inline ADAL/src/urlprotocol/ADCustomHeaderHandler.m]: do we need dealloc function in this class to release objects like s_customHeadersForSingleUse?

File: ADAL/src/ADAuthenticationContext.m (hunk 1 above)
Comment (RPangrle) [inline ADAL/src/ADAuthenticationContext.m]: Once we pass it into the authentication context it takes ownership of it and we don't need it anymore so we release it.

File: ADAL/src/ADInstanceDiscovery.m (hunk 3 above)
Comment (RPangrle) [inline ADAL/src/ADInstanceDiscovery.m]: lowercaseString doesn't start with copy, alloc, new or mutableCopy so we don't own the reference to it. Similarly we don't own objects passed to us in parameters

File: ADAL/src/urlprotocol/ADCustomHeaderHandler.m (hunk 5 above)
Comment (RPangrle) [inline ADAL/src/urlprotocol/ADCustomHeaderHandler.m]: File statics live for the life time of the app. There's no particular point releasing "on quit" when the OS is just going to reclaim all the memory anyways, you'll only slow down app quit.

File: ADAL/src/ADAuthenticationContext.m (hunk 1 above)
Comment (RPangrle) [inline ADAL/src/ADAuthenticationContext.m]: Sorry, wrong answer. In this case we didn't call alloc (someone else did) so we don't have a retain count to release or autorelease.

Comment (jasoncoolmax) [inline ADAL/src/ADAuthenticationContext.m]: Got it :)

File: ADAL/src/ADInstanceDiscovery.m (hunk 3 above)
Comment (jasoncoolmax) [inline ADAL/src/ADInstanceDiscovery.m]: Got it! :)

File: ADAL/src/urlprotocol/ADCustomHeaderHandler.m (hunk 5 above)
Comment (jasoncoolmax) [inline ADAL/src/urlprotocol/ADCustomHeaderHandler.m]: I see, make sense :)

**Comment by jasoncoolmax** [comment]
I signed off :)

File: ADAL/src/ADAuthenticationSettings.m [hunk 6]
```diff
@@ -15,11 +15,17 @@
...
 #import "ADAuthenticationSettings.h"
-#import "ADKeychainTokenCache.h"
 
 @implementation ADAuthenticationSettings
 
+@synthesize requestTimeOut = _requestTimeOut;
+@synthesize expirationBuffer = _expirationBuffer;
+
+#if !TARGET_OS_IPHONE
+@synthesize defaultCacheDelegate = _defaultCacheDelegate;
```
Comment (kpanwar) [inline ADAL/src/ADAuthenticationSettings.m]: Consider renaming to include 'storage' to clear the purpose.

File: ADAL/src/ADInstanceDiscovery.m [hunk 7]
```diff
@@ -174,43 +163,97 @@ -(void) validateAuthority: (NSString*) authority
...
     AD_LOG_VERBOSE(@"Authority Validation Cache", nil, message);
+    return YES;
+}
+
+- (ADAuthenticationError *)processWebReponse:(ADWebResponse *)webResponse
+                               authorityHost:(NSString *)authorityHost
+                               correlationId:(NSUUID *)correlationId
+{
+    NSInteger code = webResponse.statusCode;
+    if (!(code == 200 || code == 400 || code == 401))
```
Comment (kpanwar) [inline ADAL/src/ADInstanceDiscovery.m]: We should  check if error 400 returns a JSON. If it does not then deserialization would fail

Comment (RPangrle) [inline ADAL/src/ADInstanceDiscovery.m]: Um, we are... This if statement is if the code is not (200 or 400 or 401)...

File: ADAL/src/ADAuthenticationSettings.m (hunk 6 above)
Comment (RPangrle) [inline ADAL/src/ADAuthenticationSettings.m]: :+1:

File: ADAL/src/public/ADTokenCacheItem.h [hunk 8]
```diff
@@ -25,6 +25,20 @@
     Objects of this class are used in the key-based token cache store.
     See the key extraction function for details on how the keys are constructed. */
 @interface ADTokenCacheItem : NSObject<NSCopying , NSSecureCoding>
+{
+    NSUInteger _hash;
+    NSString* _resource;
+    NSString* _authority;
+    NSString* _clientId;
+    NSString* _familyId;
```
Comment (kpanwar) [inline ADAL/src/public/ADTokenCacheItem.h]: where is familyId being set? also, we are using "foci" as a value in the server response.

**Comment by kpanwar** [comment]
Sign off with comments and an issue filed. https://github.com/AzureAD/azure-activedirectory-library-for-objc/issues/498

File: ADAL/src/public/ADTokenCacheItem.h (hunk 8 above)
Comment (RPangrle) [inline ADAL/src/public/ADTokenCacheItem.h]: The family ID key has changed several times now. That's a separate issue to be handled in another PR.

**Comment by RPangrle** [comment]
#479
